proving.go

```diff
@@ -0,0 +1,232 @@
+package merkle
+
+import (
+	"errors"
+	"math"
+)
+
+const NodeSize = 32
+
+type NodeReader interface {
+	Seek(index uint64) error
+	ReadNext() ([]byte, error)
+	Width() uint64
+}
+
+func GenerateProof(
+	provenLeafIndices []uint64,
+	readers map[uint]NodeReader,
```

Comment:
There is quite a lot of logic around accessing and maintaining the cache which justifies building a dedicated cache struct with supporting methods